subscription.go

```diff
@@ -31,3 +52,65 @@ func (sub *Subscription) Next(ctx context.Context) (*Message, error) {
 func (sub *Subscription) Cancel() {
 	sub.cancelCh <- sub
 }
+
+func (sub *Subscription) close() {
+	close(sub.ch)
+}
+
+func (sub *Subscription) sendNotification(evt PeerEvent) {
+	sub.backlogMx.Lock()
+	defer sub.backlogMx.Unlock()
+
+	sub.addToBacklog(evt)
+
+	select {
+	case sub.backlogCh <- struct{}{}:
+	default:
+	}
+}
+
+// addToBacklog assumes a lock has been taken to protect the backlog
+func (sub *Subscription) addToBacklog(evt PeerEvent) {
+	e, ok := sub.evtBacklog[evt.Peer]
+	if !ok {
+		sub.evtBacklog[evt.Peer] = evt.Type
+	} else if e != evt.Type {
+		delete(sub.evtBacklog, evt.Peer)
+	}
+}
+
+// pullFromBacklog assumes a lock has been taken to protect the backlog
+func (sub *Subscription) pullFromBacklog() (PeerEvent, bool) {
+	for k, v := range sub.evtBacklog {
+		evt := PeerEvent{Peer: k, Type: v}
+		delete(sub.evtBacklog, k)
+		return evt, true
+	}
+	return PeerEvent{}, false
+}
+
+// NextPeerEvent returns the next event regarding subscribed peers
+// Guarantees: Peer Join and Peer Leave events for a given peer will fire in order.
+// Unless a peer both Joins and Leaves before NextPeerEvent emits either event
+// all events will eventually be received from NextPeerEvent.
+func (sub *Subscription) NextPeerEvent(ctx context.Context) (PeerEvent, error) {
+	sub.nextEventMx.Lock()
```

Comment:
Ok, so, I was wrong. This has a problem:

1. Thread 1: Call `NextPeerEvent(ctxa)`
2. Thread 1: Take lock.
3. Thread 2: Call `NextPeerEvent(ctxb)`
4. Thread 3: Cancel `ctxb`.

I think we need to go the loop route where any thread taking from the backlog puts the flag back in the backlog.

Reply:
right, if you cancel because we're using mutexes and not channels nothing's going to happen. I guess we can always keep putting the flag in.

However, it does mean that if for some reason a consumer spun up two goroutines to read events then a Join and a Leave could come out of order. Do we care about this case?

Reply:
No, that's their problem. I mean, we could just say "only do this on one thread" but... I don't like weird contracts like that.